cpp/src/arrow/compute/kernels/vector_swizzle.cc

```diff
@@ -0,0 +1,415 @@
+
+  Status AllocateValidityBufAndFill(bool valid) {
+    DCHECK_EQ(validity_buf, nullptr);
+
+    ARROW_ASSIGN_OR_RAISE(validity_buf, ctx->AllocateBitmap(output_length));
+    auto validity = validity_buf->mutable_data_as<uint8_t>();
+    std::memset(validity, valid ? 0xff : 0, validity_buf->size());
+
+    return Status::OK();
+  }
+
+  Status AllocateDataBuf(const DataType& output_type) {
+    DCHECK_EQ(data_buf, nullptr);
+
+    ARROW_ASSIGN_OR_RAISE(data_buf,
+                          ctx->Allocate(output_length * output_type.byte_width()));
+
+    return Status::OK();
+  }
+
+  template <typename Type, typename OutputCType = typename Type::c_type>
+  Status AllocateDataBufAndFill(const Type& output_type, OutputCType value) {
+    RETURN_NOT_OK(AllocateDataBuf(output_type));
+
+    OutputCType* data = data_buf->mutable_data_as<OutputCType>();
+    for (int64_t i = 0; i < output_length; ++i) {
+      data[i] = value;
+    }
+
+    return Status::OK();
+  }
+
+  template <typename Type, bool likely_many_nulls>
+  Status Execute() {
+    using OutputCType = typename Type::c_type;
+
+    uint8_t* validity = nullptr;
+    if constexpr (likely_many_nulls) {
+      DCHECK_NE(validity_buf, nullptr);
+      validity = validity_buf->mutable_data_as<uint8_t>();
+    } else {
+      DCHECK_EQ(validity_buf, nullptr);
+    }
+    DCHECK_NE(data_buf, nullptr);
+    OutputCType* data = data_buf->mutable_data_as<OutputCType>();
+    int64_t inverse = 0;
+    RETURN_NOT_OK(ExecType::template VisitIndices(
+        indices,
+        [&](IndexCType index) {
+          if (ARROW_PREDICT_TRUE(index >= 0 &&
+                                 static_cast<int64_t>(index) < output_length)) {
+            data[index] = static_cast<OutputCType>(inverse);
+            // If many nulls, set validity to true for valid values.
+            if constexpr (likely_many_nulls) {
+              bit_util::SetBitTo(validity, index, true);
+            }
+          }
+          ++inverse;
+          return Status::OK();
+        },
+        [&]() {
+          ++inverse;
+          return Status::OK();
+        }));
+
+    // If not many nulls, run another pass iterating over the data to set the validity
+    // to false if the value is "impossible". The validity buffer is on demand allocated
+    // and initialized all-true when the first "impossible" value is seen.
+    if constexpr (!likely_many_nulls) {
+      for (int64_t i = 0; i < output_length; ++i) {
+        if (ARROW_PREDICT_FALSE(data[i] == static_cast<OutputCType>(input_length))) {
+          if (ARROW_PREDICT_FALSE(!validity_buf)) {
+            RETURN_NOT_OK(AllocateValidityBufAndFill(true));
+            validity = validity_buf->mutable_data_as<uint8_t>();
+          }
+          bit_util::SetBitTo(validity, i, false);
+        }
+      }
+    }
+
+    return Status::OK();
+  }
+
+  template <typename VISITOR, typename... ARGS>
+  friend Status arrow::VisitTypeInline(const DataType&, VISITOR*, ARGS&&... args);
```

Comment:
Or perhaps we can remove this and make the required methods public? (this is a module-private class in any case)